### PR TITLE
Adding UIA support for ScrollBar

### LINKED ISSUE #4983: ScrollBar control does not support UIA accessibility
**Environment**
- Microsoft Windows [Version 10.0.19041.388]
- .NET Core SDK 6.0.100-preview.3.21202.5

**Have you experienced this same bug with .NET Framework?:**
- Yes

**Problem description**
image

Accessibility providers:
`[pid:10384,providerId:0x811196 Main:Nested [pid:10832,providerId:0x811196 Main(parent link):Microsoft: MSAA Proxy (unmanaged:UIAutomationCore.dll)]; Hwnd(parent link):Microsoft: HWND Proxy (unmanaged:uiautomationcore.dll)]`

**Expected behavior:**
There should be UIA accessibility provider from WinForms.

**Minimal repro:**
Create WinForms app with TrackBar control and analyze it using Inspect.

**Related item:**
This work item is the sub of #3037

### PR DESCRIPTION
**Pull request by SergeySmirnov-Akvelon**
Fixes #4983

## Proposed changes
- Updated "SupportsUiaProviders" flag.
- Added and implemented ScrollBar accessible objects
- Added unit tests

## Customer Impact
**Before fix**:
image

Accessibility providers:
`[pid:10384,providerId:0x811196 Main:Nested [pid:10832,providerId:0x811196 Main(parent link):Microsoft: MSAA Proxy (unmanaged:UIAutomationCore.dll)]; Hwnd(parent link):Microsoft: HWND Proxy (unmanaged:uiautomationcore.dll)]`

**Afrer fix:**
image
Accessibility providers:
`[pid:30620,providerId:0xD1472 Main:Nested [pid:52468,providerId:0xD1472 Main(parent link):Unidentified Provider (unmanaged:coreclr.dll)]; Hwnd(parent link):Microsoft: HWND Proxy (unmanaged:uiautomationcore.dll)]`

## Regression? 

- No

## Risk
- Minimal 

## Test methodology 
- Unit tests 
- CTI team

## Accessibility testing  
- Narrator
- Inspect
- Accessibility Insights

## Test environment(s) 
- Microsoft Windows [Version 10.0.19041.388]
- .NET Core SDK: 6.0.100-preview.3.21202.5

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/5044)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs [hunk 1]
```diff
@@ -15,25 +16,141 @@ public partial class ScrollBar
...
         internal class ScrollBarAccessibleObject : ControlAccessibleObject
         {
-            private ScrollBar _owningScrollBar;
+            private protected readonly ScrollBar _owningScrollBar;
+            private ScrollBarFirstLineButtonAccessibleObject? _firstLineButtonAccessibleObject;
+            private ScrollBarFirstPageButtonAccessibleObject? _firstPageButtonAccessibleObject;
+            private ScrollBarLastLineButtonAccessibleObject? _lastLineButtonAccessibleObject;
+            private ScrollBarLastPageButtonAccessibleObject? _lastPageButtonAccessibleObject;
+            private ScrollBarThumbAccessibleObject? _thumbAccessibleObject;
+            internal ScrollBarAccessibleObject(ScrollBar owningScrollBar) : base(owningScrollBar)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs:25]: Missing blank line

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs [hunk 2]
```diff
@@ -15,25 +16,141 @@ public partial class ScrollBar
     {
         internal class ScrollBarAccessibleObject : ControlAccessibleObject
         {
-            private ScrollBar _owningScrollBar;
+            private protected readonly ScrollBar _owningScrollBar;
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs:19]: No `protected` fields, turn to a property

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs [hunk 3]
```diff
@@ -15,25 +16,141 @@ public partial class ScrollBar
...
+            private ScrollBarLastLineButtonAccessibleObject? _lastLineButtonAccessibleObject;
+            private ScrollBarLastPageButtonAccessibleObject? _lastPageButtonAccessibleObject;
+            private ScrollBarThumbAccessibleObject? _thumbAccessibleObject;
+            internal ScrollBarAccessibleObject(ScrollBar owningScrollBar) : base(owningScrollBar)
+            {
+                _owningScrollBar = owningScrollBar;
+            }
+
+            internal ScrollBarFirstLineButtonAccessibleObject FirstLineButtonAccessibleObject
+                            => _firstLineButtonAccessibleObject ??= new ScrollBarFirstLineButtonAccessibleObject(_owningScrollBar);
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs:31]: Incorrect indentation

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs [hunk 4]
```diff
@@ -15,25 +16,141 @@ public partial class ScrollBar
...
+                    var runtimeId = new int[2];
+                    runtimeId[0] = RuntimeIDFirstItem;
+                    runtimeId[1] = Interop.PARAM.ToInt(_owningScrollBar.InternalHandle);
+                    return runtimeId;
+                }
             }
 
-            internal override bool IsIAccessibleExSupported() => true;
+            internal ScrollBarThumbAccessibleObject ThumbAccessibleObject
+                            => _thumbAccessibleObject ??= new ScrollBarThumbAccessibleObject(_owningScrollBar);
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarAccessibleObject.cs:61]: ditto

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+using System.Drawing;
+using static Interop;
+
+namespace System.Windows.Forms
+{
+    public partial class ScrollBar
+    {
+        internal class ScrollBarChildAccessibleObject : AccessibleObject
+        {
+            private protected readonly ScrollBar _owningScrollBar;
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs:14]: ditto

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+            public override string? Name
+                => ParentInternal.GetSystemIAccessibleInternal()?.get_accName(GetChildId());
+
+            public override AccessibleRole Role
+            {
+                get
+                {
+                    var accRole = ParentInternal.GetSystemIAccessibleInternal()?.get_accRole(GetChildId());
+                    return accRole is not null
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs:51]: 1. Use explicit type
2. Use pattern matching

**Review by RussKie — COMMENTED** [review]
Similar comments as for https://github.com/dotnet/winforms/pull/4939

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+                    {
+                        return Rectangle.Empty;
+                    }
+
+                    int left = 0;
+                    int top = 0;
+                    int width = 0;
+                    int height = 0;
+
+                    // The "GetChildId" method returns to the id of the trackbar element,
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs:35]: Throughout this file you say `trackbar`, I think you actually mean `scrollbar child`.

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs:35]: And perhaps it's sufficient to have this comment only for the first occurrence where we query the internal  accessible object?

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            public override AccessibleStates State
+                // The "GetChildId" method returns to the id of the trackbar element,
+                // which allows to use the native "get_accState" method to get the "State" property
+                => ParentInternal.GetSystemIAccessibleInternal()?.get_accState(GetChildId()) is object accState
+                    ? (AccessibleStates)accState
+                    : AccessibleStates.None;
+
+            internal override UiaCore.IRawElementProviderFragmentRoot FragmentRoot => ParentInternal;
+
+            internal virtual bool IsDisplayed => true;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs:74]: Was this always true previously? I think this property should be checking if the owner scroll bar is visible.

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+                    _ => base.FragmentNavigate(direction)
+                };
+
+            internal override object? GetPropertyValue(UiaCore.UIA propertyID)
+                => propertyID switch
+                {
+                    UiaCore.UIA.RuntimeIdPropertyId => RuntimeId,
+                    UiaCore.UIA.AccessKeyPropertyId => string.Empty,
+                    UiaCore.UIA.ControlTypePropertyId => UiaCore.UIA.ButtonControlTypeId,
+                    UiaCore.UIA.IsEnabledPropertyId => OwningScrollBar.Enabled && IsDisplayed,
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs:109]: Is `IsDisplayed` property needed here? If this child element is not visible, then it's UIA `enabled` property is set to false, which is not entirely accurate. The user would think that they need to make the UI element "enabled" while in fact they need to scroll in the parent UI to make the scroll bar element visible.

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarFirstLineButtonAccessibleObject.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+{
+    public partial class ScrollBar
+    {
+        internal class ScrollBarFirstLineButtonAccessibleObject : ScrollBarChildAccessibleObject
+        {
+            public ScrollBarFirstLineButtonAccessibleObject(ScrollBar owningScrollBar) : base(owningScrollBar)
+            {
+            }
+
+            internal override Interop.UiaCore.IRawElementProviderFragment? FragmentNavigate(Interop.UiaCore.NavigateDirection direction)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarFirstLineButtonAccessibleObject.cs:15]: Consider adding a using for Interop.

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarFirstPageButtonAccessibleObject.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+            internal override Interop.UiaCore.IRawElementProviderFragment? FragmentNavigate(Interop.UiaCore.NavigateDirection direction)
+            {
+                if (!OwningScrollBar.IsHandleCreated || !IsDisplayed)
+                {
+                    return null;
+                }
+
+                return direction switch
+                {
+                    Interop.UiaCore.NavigateDirection.PreviousSibling => ParentInternal.FirstLineButtonAccessibleObject,
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarFirstPageButtonAccessibleObject.cs:34]: `using static Interop;`?

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarThumbAccessibleObject.cs [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+            internal override Interop.UiaCore.IRawElementProviderFragment? FragmentNavigate(Interop.UiaCore.NavigateDirection direction)
+            {
+                if (!OwningScrollBar.IsHandleCreated)
+                {
+                    return null;
+                }
+
+                return direction switch
+                {
+                    Interop.UiaCore.NavigateDirection.PreviousSibling
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarThumbAccessibleObject.cs:28]: remove `Interop` namespace name, you have `using static Interop;` in this file.

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarLastPageButtonAccessibleObject.cs [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+            internal override Interop.UiaCore.IRawElementProviderFragment? FragmentNavigate(Interop.UiaCore.NavigateDirection direction)
+            {
+                if (!OwningScrollBar.IsHandleCreated || !IsDisplayed)
+                {
+                    return null;
+                }
+
+                return direction switch
+                {
+                    Interop.UiaCore.NavigateDirection.PreviousSibling => ParentInternal.ThumbAccessibleObject,
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarLastPageButtonAccessibleObject.cs:35]: add `using static Interop;`

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarLastLineButtonAccessibleObject.cs [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+            {
+                if (!OwningScrollBar.IsHandleCreated)
+                {
+                    return null;
+                }
+
+                return direction switch
+                {
+                    Interop.UiaCore.NavigateDirection.NextSibling => null,
+                    Interop.UiaCore.NavigateDirection.PreviousSibling
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarLastLineButtonAccessibleObject.cs:25]: add `using static Interop;`

**Review by Tanya-Solyanik — COMMENTED** [review]
added some questions about `Enabled` UIA property.

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs (hunk 7 above)
Comment (SergeySmirnov-Akvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs]: Fixed typos. Removed unneeded comments

File: src/System.Windows.Forms/src/PublicAPI.Shipped.txt [hunk 15]
```diff
@@ -4349,7 +4349,6 @@ override System.Windows.Forms.GroupBox.DisplayRectangle.get -> System.Drawing.Re
 override System.Windows.Forms.GroupBox.ProcessMnemonic(char charCode) -> bool
 override System.Windows.Forms.GroupBox.ScaleControl(System.Drawing.SizeF factor, System.Windows.Forms.BoundsSpecified specified) -> void
 override System.Windows.Forms.GroupBox.WndProc(ref System.Windows.Forms.Message m) -> void
-override System.Windows.Forms.HScrollBar.CreateAccessibilityInstance() -> System.Windows.Forms.AccessibleObject!
```
Comment (RussKie) [inline src/System.Windows.Forms/src/PublicAPI.Shipped.txt:4352]: ❗ We can't delete entries from the Shipped file. You need to add matching entries into the Unshipped: https://github.com/dotnet/roslyn/blob/main/docs/Adding%20Optional%20Parameters%20in%20Public%20API.md#note

**Review by RussKie — COMMENTED** [review]
🔥

**Comment by RussKie** [comment]
Forgot to clean or push?
image

Also make sure it runs locally:
```
    System.Windows.Forms.Maui.IntegrationTests.WinformsMauiVScrollBarTests.MauiVScrollBarTests [FAIL]
      System.IO.FileNotFoundException : File does not exist
      Stack Trace:
        /_/src/System.Windows.Forms/tests/IntegrationTests/System.Windows.Forms.IntegrationTests.Common/TestHelpers.cs(53,0): at System.Windows.Forms.IntegrationTests.Common.TestHelpers.GetExePath(String projectName)
        /_/src/System.Windows.Forms/tests/IntegrationTests/System.Windows.Forms.Maui.IntegrationTests/MauiTestHelper.cs(40,0): at System.Windows.Forms.Maui.IntegrationTests.MauiTestHelper.GetScenarios(String projectName)
        /_/src/System.Windows.Forms/tests/IntegrationTests/System.Windows.Forms.Maui.IntegrationTests/MauiDataAttribute.cs(55,0): at System.Windows.Forms.Maui.IntegrationTests.MauiDataAttribute.GetData(MethodInfo testMethod)
```

**Comment by Tanya-Solyanik** [comment]
@DmitryGorokhov - see the before/after screenshots and red/green highlights to see when MSAA proxy is used and when our custom providers are

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+                    UiaCore.UIA.IsPasswordPropertyId => false,
+                    UiaCore.UIA.IsOffscreenPropertyId => (State & AccessibleStates.Offscreen) == AccessibleStates.Offscreen,
+                    UiaCore.UIA.HasKeyboardFocusPropertyId => false,
+                    UiaCore.UIA.IsKeyboardFocusablePropertyId => false,
+                    _ => base.GetPropertyValue(propertyID)
+                };
+
+            internal override void Invoke()
+            {
+                if (OwningScrollBar.IsHandleCreated && IsDisplayed)
```
[the remark this answers is not in the document]
Comment (SergeySmirnov-Akvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarChildAccessibleObject.cs:111]: In this case, the `IsDisplayed` flag, in addition to the fact that the element is invisible, means that the trackbar already has a minimum/maximum value and calling the `Invoke` method for element will not lead to any action.

File: src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarFirstPageButtonAccessibleObject.cs [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+                    return OwningScrollBar._scrollOrientation == ScrollOrientation.HorizontalScroll
+                        && OwningScrollBar.RightToLeft == RightToLeft.Yes
+                            ? ParentInternal.UIMaximum > OwningScrollBar.Value
+                            : OwningScrollBar.Minimum != OwningScrollBar.Value;
+                }
+            }
+
+            internal override IRawElementProviderFragment? FragmentNavigate(NavigateDirection direction)
+            {
+                if (!OwningScrollBar.IsHandleCreated || !IsDisplayed)
```
[the remark this answers is not in the document]
Comment (SergeySmirnov-Akvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/ScrollBar.ScrollBarFirstPageButtonAccessibleObject.cs]: I have investigated this case and found that if the element is invisible, than `FragmentNavigate(NavigateDirection.Parent)` method returns `Parent` correctly, but `FragmentNavigate(NavigateDirection.NextSibling)` and `FragmentNavigate(NavigateDirection.PreviousSibling)` methods return `null` value.

I fixed this case for `ScrollBar` and updated the tests. I also fixed this case for `TrackBar`.

The `IsDisplayed` property means that the `ScrollBar` has a minimum / maximum value. Therefore this element is invisible and does not require user interaction.

**Comment by dreddy-work** [comment]
@Olina-Zhang

**Comment by SergeySmirnov-Akvelon** [comment]
Approved by CTI team

**Review by RussKie — APPROVED** [review]

**Pull request merged**